README.md
# Logardian

Inspired by NestJS Logger, Logardian was built to output minimalistic, readable logs.

## Roadmap

- [x] OTEL Support
- [x] NodeJS v16.17.1
- [ ] Logging to file

## Installation

> Note: Logardian with version 3.0.0 or higher requires NodeJS 16+ because of async hooks. If you don't want them downgrade to version 2.1.0

```bash
npm i --save logardian
```
    
## Features

- Various layers of logs that can be turned on/off via config
- OpenTelemetry trace ID support
- In production mode debug() does not work
- Datetime in UTC format
- Format of logs: `[time] [level] [layer] [message]`
- In debug mode the path and name of the function that called the log is displayed
- Can be used instead of NestJS Logger
- Can log any objects, arrays, variables
- Modes: normal or json format output
- Colors!

  
## Usage/Examples

```ts
import { Logardian } from 'logardian'

const logger = new Logardian()

logger.configure({
    trace: false,
    labels: ['users', '*.debug']
})

logger.log(`Hi! I'm info log example`)
logger.warn(`Hi! I'm warn log example`)
logger.error(`Hi! I'm error log example`)
logger.verbose(`Hi! I'm verbose log example`)
logger.debug({ some: 'object' })

logger.markTime('marker')

setTimeout(() => {
    logger.measureTime('marker', 'Function take {n} ms to execute')
}, 2000)

logger.log(`I will log`, { label: 'users' })
logger.log(`I will log too`, { label: 'auth.debug' })
logger.log(`I will not log :(`, { label: 'database' })
```

### Default output

![](./images/logs.png)


### Json output

```ts
logger.configure({
    json: true
})
```

```bash
{"timestamp":"2022-10-10T12:39:40.012Z","message":"Starting Nest application...","level":"log"}
{"timestamp":"2022-10-10T12:39:40.017Z","message":"AppModule dependencies initialized","level":"log"}
{"timestamp":"2022-10-10T12:39:40.020Z","message":"Nest application successfully started","level":"log"}
{"timestamp":"2022-10-10T12:39:40.022Z","message":"Hi! I'm info log example","level":"log"}
{"timestamp":"2022-10-10T12:39:40.022Z","message":"Hi! I'm warn log example","level":"warn"}
{"timestamp":"2022-10-10T12:39:40.022Z","message":"Hi! I'm error log example","level":"error"}
{"timestamp":"2022-10-10T12:39:40.022Z","message":"Hi! I'm verbose log example","level":"verbose"}
{"timestamp":"2022-10-10T12:39:40.022Z","message":"{\"some\":\"object\"}","level":"debug"}
{"timestamp":"2022-10-10T12:39:40.023Z","message":"I will log","level":"log","label":"users"}
{"timestamp":"2022-10-10T12:39:40.024Z","message":"I will log too","level":"log","label":"auth.debug"}
{"timestamp":"2022-10-10T12:39:42.024Z","message":"Function take 2002.041 ms to execute","level":"timer"}
```

### Labels

Labels now support glob patterns! You can dynamically enable and disable the logs you need via `logger.configure()`. For example:

```ts
import { Logardian } from 'logardian'

const logger = new Logardian()

logger.configure({
    labels: ['users.*']
})

logger.log('User sent mail', { label: 'users.email' }) // will log
logger.log('User registered', { label: 'users.auth.registration' }) // will log
logger.log('User authorized', { label: 'users.auth.authorization' }) // will log
logger.log('Database connected', { label: 'database' }) // will NOT log
logger.log('User entity created', { label: 'database.users' }) // will NOT log
```

### OTEL Trace IDs

Use trace ID from your OTEL

```ts
// your create user logic

logger.log('User has been created')
// [2022-10-05 11:34:41.621] [7a22fdae427ddd12ace3a129e344121b]    LOG: User has been created
//                                    ^ unique trace id

// your send email for user logic here

logger.log('Email for user was sent')
// [2022-10-05 11:34:47.317] [7a22fdae427ddd12ace3a129e344121b]    LOG: Mail for user was sent
//                                    ^ same trace id

const traceId = logger.getTraceId()
// 7a22fdae427ddd12ace3a129e344121b
```

### Colors

If you don't like the set of colors logardian provides you can change them in `configure()` function.

```ts
const logger = new Logardian()

logger.configure({
    colors: {
        timestamp: '#ABC'
        traceId: '#CCC
        label: '#FFFFFF
        message: '#FFFFFF
        trace: '#FFFFFF
        stack: '#000000
    }
})
```

## Environment Variables

`NODE_ENV` production start does not show debug() logs

## FAQ

#### How does it implement NestJS Logger without any framework libs?

We made logger based on [LoggerService](https://github.com/nestjs/nest/blob/master/packages/common/services/logger.service.ts) but we don't explicitly import it so that we stay dependless of NestJS libraries. But you can also use the Logardian instead of common NestJS logger.

```ts
// main.ts
import { Logardian } from 'logardian'

const logger = new Logardian()

async function bootstrap(): Promise<void> {
    const app = await NestFactory.create(AppModule, { logger })

    await app.listen(port, hostname, () =>
        logger.log(`Server running at ${hostname}:${port}`),
    )
}
```

#### How can I use logardian in my NestJS service?

Simply create a new logger class

```ts
import { Logardian } from 'logardian'

@Injectable()
export class CatService {
    private readonly _logger = new Logardian()
}
```

#### I do not see my logs with label

Specify labels you want to log or write `*` to log every log with label. 
Working in production and development mode

Logardian is a singleton, so it means that `configure()` works on all Logardian instances

```ts
import { Logardian } from 'logardian'

const logger = new Logardian()

logger.configure({
    labels: '*', // or ['database', 'events'] or false
    trace: false,
    json: true,
    traceId: true
})
```

#### I do not want to see caller and path. How can I turn off them globally?

Specify 'false' on logardian config. If you specify `trace: true` in logger function trace will log in spite of config option

Priority of trace from high to low:

1. Production mode
2. `logger.log('Hello', { trace: true })`
3. `logger.configure({ trace: false })`



## License


[MIT](https://github.com/i-link-pro-team/logardian/blob/main/LICENSE)

  

CHANGELOG.md
# [4.0.0](https://github.com/temarusanov/logardian/compare/v3.2.0...v4.0.0) (2022-10-21)


### Bug Fixes

* otel to devDeps for npm ci ([d18594f](https://github.com/temarusanov/logardian/commit/d18594fcafc8d66393e408dfc7921bfdf1df89d7))
* remove console.logs ([1744734](https://github.com/temarusanov/logardian/commit/1744734afdcc7fe54131f0482802404e37d168fb))


* Merge pull request #33 from temarusanov/fix/trace-id ([b24970a](https://github.com/temarusanov/logardian/commit/b24970ab522bbcba60ae65267d71d9802c9be2ba)), closes [#33](https://github.com/temarusanov/logardian/issues/33)


### Features

* otel trace ids ([b5f4bb2](https://github.com/temarusanov/logardian/commit/b5f4bb20ad9c9c4da431d6b7509933f290a433bc))


### BREAKING CHANGES

* now with OTEL support

# [3.2.0](https://github.com/temarusanov/logardian/compare/v3.1.0...v3.2.0) (2022-10-13)


### Features

* add getTraceId function ([b05d541](https://github.com/temarusanov/logardian/commit/b05d541c74a9557f9d8c5c15273c10915ceed4f3))

# [3.1.0](https://github.com/temarusanov/logardian/compare/v3.0.0...v3.1.0) (2022-10-06)


### Features

* add color config, redesign logs ([a7ddf62](https://github.com/temarusanov/logardian/commit/a7ddf626f59fe1510751db1d2ce55d0dec41d719))
* add glob pattern for labels ([bd5ba00](https://github.com/temarusanov/logardian/commit/bd5ba00b05b51ed7891c72de2c54caca353f844e))

# [3.0.0](https://github.com/temarusanov/logardian/compare/v2.1.0...v3.0.0) (2022-10-05)


### Features

* add async hooks and trace id ([b0be1ca](https://github.com/temarusanov/logardian/commit/b0be1ca4c6d2a51fd4273776e4d68db5cb226810))
* remove json log space indentation ([eb8e5e8](https://github.com/temarusanov/logardian/commit/eb8e5e82daeb9181c6aea99cc2721cf4bbe3c990))
* update deps ([a2cd79f](https://github.com/temarusanov/logardian/commit/a2cd79fa6b8bab17fefd1d80140550f485309a2b))


### BREAKING CHANGES

* move to nodejs 16.17.1

# [2.1.0](https://github.com/temarusanov/logardian/compare/v2.0.2...v2.1.0) (2022-09-27)


### Features

* add timer feature ([0b70818](https://github.com/temarusanov/logardian/commit/0b7081884fcb43b79184d27018ff7ef0cc14a87e))

## [2.0.2](https://github.com/temarusanov/logardian/compare/v2.0.1...v2.0.2) (2021-12-17)


### Bug Fixes

* add stacktrace regex ([6a28df1](https://github.com/temarusanov/logardian/commit/6a28df110556415bee5fca62cdd091b32ac68f05))

## [2.0.1](https://github.com/temarusanov/logardian/compare/v2.0.0...v2.0.1) (2021-11-05)


### Bug Fixes

* function trace off priority ([cb93fee](https://github.com/temarusanov/logardian/commit/cb93feea897a6c4f6e13cb03a2bd7cc17a03a30c))

# [2.0.0](https://github.com/temarusanov/logardian/compare/v1.4.0...v2.0.0) (2021-11-05)


* Merge pull request #27 from temarusanov/feat/improvements ([ad409e7](https://github.com/temarusanov/logardian/commit/ad409e7674fccf17ff47c4450e4e71d99c0cf6b3)), closes [#27](https://github.com/temarusanov/logardian/issues/27)


### Features

* **refactor:** migrate from .env settings to incode config ([8930e4f](https://github.com/temarusanov/logardian/commit/8930e4f3844987e9cd6a183a47b70f04f09bf6fc))


### BREAKING CHANGES

* .env config will not work in this version

# [1.4.0](https://github.com/temarusanov/logardian/compare/v1.3.2...v1.4.0) (2021-11-05)


### Features

* prod mode dont show json logs, new env setting now ([a87797f](https://github.com/temarusanov/logardian/commit/a87797facf5c67b14c3b222ee5c2ec0793b3a04b))

## [1.3.2](https://github.com/temarusanov/logardian/compare/v1.3.1...v1.3.2) (2021-11-04)


### Bug Fixes

* options undefined in error log ([ab1a06e](https://github.com/temarusanov/logardian/commit/ab1a06ee4a43911f817a937f0636fe5474ee0e09))

## [1.3.1](https://github.com/temarusanov/logardian/compare/v1.3.0...v1.3.1) (2021-11-04)


### Bug Fixes

* stack undefined check ([a0a72f5](https://github.com/temarusanov/logardian/commit/a0a72f5c1d9a2f43550b2ebdab741eda7e8bd4c7))

# [1.3.0](https://github.com/temarusanov/logardian/compare/v1.2.0...v1.3.0) (2021-11-04)


### Features

* **refactor:** fully abandon the winston to stdout, using nestjs logger as a base ([2940b2e](https://github.com/temarusanov/logardian/commit/2940b2e280e902cd689908b72257f51ac2991eb9))

# [1.2.0](https://github.com/i-link-pro-team/logardian/compare/v1.1.1...v1.2.0) (2021-09-30)


### Bug Fixes

* caller and trace shows in prod mode ([49159ef](https://github.com/i-link-pro-team/logardian/commit/49159ef97766157ba0c20969401e1ee6210cbf58))
* function caller error ([5f4fa6b](https://github.com/i-link-pro-team/logardian/commit/5f4fa6b36f691b541f706ae286d63a9fa0d0e0f5))


### Features

* add docs to prod mode ([aec726b](https://github.com/i-link-pro-team/logardian/commit/aec726bc2b6491f37d74de9d744579aef26212a8))
* add docs, rebase ([af76aaa](https://github.com/i-link-pro-team/logardian/commit/af76aaa9db47840f95b4a24f3c6581786ee3f640))

## [1.1.1](https://github.com/i-link-pro-team/logardian/compare/v1.1.0...v1.1.1) (2021-09-24)


### Bug Fixes

* add milliseconds to date log ([dae933d](https://github.com/i-link-pro-team/logardian/commit/dae933dbf44e42927fc24a6a9c2706e67fa1dc2a))
* production log didnt show metadata ([9d70071](https://github.com/i-link-pro-team/logardian/commit/9d70071497ad0bbb262885f14a50a9c86172fc55))

# [1.1.0](https://github.com/i-link-pro-team/logardian/compare/v1.0.2...v1.1.0) (2021-09-23)


### Features

* templates for issue ([8816803](https://github.com/i-link-pro-team/logardian/commit/881680303c920324938d5a9bda14aa3ea34e4091))

## [1.0.2](https://github.com/i-link-pro-team/logardian/compare/v1.0.1...v1.0.2) (2021-09-23)


### Bug Fixes

* remove dateformat, use common date ([a3f2381](https://github.com/i-link-pro-team/logardian/commit/a3f2381f9567faba847a7cc2ee34fad8836fad01))

## [1.0.1](https://github.com/i-link-pro-team/logardian/compare/v1.0.0...v1.0.1) (2021-09-23)


### Bug Fixes


lib/logger.ts
require('dotenv').config()
import { performance } from 'perf_hooks'
import { LoggerInterface, LogLevel } from './interfaces/logger.interface'
import { clc } from './utils/cli-colors.util'
import { isPlainObject } from './utils/utils'
import { isMatch } from 'micromatch'
import * as chalk from 'chalk'
import { LogMethodOptions } from './interfaces/log-method-options.interface'
import { LoggerConfig } from './interfaces/logger-options.interface'
import { JsonLog } from './interfaces/json-log.interface'
import { DefaultLog } from './interfaces/default-log.interface'
import { context, trace } from '@opentelemetry/api'

const isProduction = process.env.NODE_ENV === 'production'

const JSON_SPACE = 2
const CONTEXT_INDEX_IN_STACK = 6

export class Logger implements LoggerInterface {
    private _timers: Map<string, number> = new Map()

    _config: LoggerConfig = {
        labels: ['*'],
        traceId: true,
        trace: false,
        colors: {
            timestamp: '#E5E5E5',
            traceId: '#E5E5E5',
            label: '#E5E5E5',
        },
    }

    log(message: any, options?: LogMethodOptions): void
    log(message: any, ...optionalParams: [...any, string?]): void
    log(message: any, ...optionalParams: any[]): void {
        const options = this._findOptions(optionalParams)
        this._printMessage(message, options, 'log')
    }

    error(message: any, options?: LogMethodOptions, stack?: string): void
    error(message: any, ...optionalParams: [...any, string?, string?]): void
    error(message: any, ...optionalParams: any[]): void {
        const { options, stack } = this._findOptionsAndStack(optionalParams)
        this._printMessage(message, options, 'error', 'stderr', stack)
    }

    warn(message: any, options?: LogMethodOptions): void
    warn(message: any, ...optionalParams: [...any, string?]): void
    warn(message: any, ...optionalParams: any[]): void {
        const options = this._findOptions(optionalParams)
        this._printMessage(message, options, 'warn')
    }

    debug(message: any, options?: LogMethodOptions): void
    debug(message: any, ...optionalParams: [...any, string?]): void
    debug(message: any, ...optionalParams: any[]): void {
        if (isProduction) {
            return
        }

        const options = this._findOptions(optionalParams)
        this._printMessage(message, options, 'debug')
    }

    verbose(message: any, options?: LogMethodOptions): void
    verbose(message: any, ...optionalParams: [...any, string?]): void
    verbose(message: any, ...optionalParams: any[]): void {
        const options = this._findOptions(optionalParams)
        this._printMessage(message, options, 'verbose')
    }

    markTime(marker: string): void {
        this._timers.set(marker, performance.now())
    }

    measureTime(
        marker: string,
        message?: string,
        options?: LogMethodOptions,
    ): number {
        const timeStart = this._timers.get(marker)

        const time = performance.now() - timeStart

        if (message) {
            const formattedMessage = message.replace(
                RegExp('{n}'),
                `${time.toFixed(3)}`,
            )

            const foundOptions = this._findOptions([options])
            this._printMessage(formattedMessage, foundOptions, 'timer')
        }

        return time
    }

    getTraceId(): string | undefined {
        try {
            const currentSpan = trace.getSpan(context.active())

            if (!currentSpan) return

            const spanContext = trace.getSpan(context.active()).spanContext()

            return spanContext.traceId
        } catch (error) {
            return
        }
    }

    private _getTimestamp(): string {
        const time = new Date().toISOString()

        // 2021-09-24T05:10:47.306Z => 2021-09-24 05:10:47
        return `[${`${time.substr(0, 10)} ${time.substr(11, 12)}`}]`
    }

    private _printMessage(
        message: any,
        options: LogMethodOptions,
        logLevel: LogLevel,
        writeStreamType?: 'stdout' | 'stderr',
        stack?: string,
    ): void {
        const { label } = options

        if (label && !this._isLabelAllowed(label)) {
            return
        }

        let computedMessage = ''
        let traceId: string | undefined

        const { json } = this._config

        if (this._config.traceId) {
            traceId = this._getTraceId(message)
        }

        if (json) {
            computedMessage = this._createJsonLog({
                message,
                label,
                level: logLevel,
                stack,
                traceId,
            })
        } else {
            const { trace } = options

            computedMessage = this._createDefaultLog({
                message,
                label,
                level: logLevel,
                stack,
                trace,
                traceId,
            })
        }

        process[writeStreamType ?? 'stdout'].write(computedMessage)
    }

    private _createJsonLog(data: JsonLog): string {
        const { message, level, label, stack, traceId } = data

        return `\n${JSON.stringify({
            timestamp: new Date().toISOString(),
            message: isPlainObject(message)
                ? JSON.stringify(message, (key, value) =>
                      typeof value === 'bigint' ? value.toString() : value,
                  )
                : (message as string),
            level,
            label,
            stack,
            traceId,
        })}`
    }

    private _createDefaultLog(data: DefaultLog): string {
        const { message, level, label, stack, trace, traceId } = data
        const { colors } = this._config

        const color = this._getColorByLogLevel(level)

        const output = isPlainObject(message)
            ? ` Object:\n${JSON.stringify(
                  message,
                  (key, value) =>
                      typeof value === 'bigint' ? value.toString() : value,
                  JSON_SPACE,
              )}`
            : ` ${message as string}`

        const timestamp = this._getTimestamp()
        const formattedTraceId = traceId !== undefined ? ` [${traceId}]` : ''

        const labelMessage = label ? ' [' + label + ']' : ''

        const formattedLogLevel = color(level.toUpperCase().padStart(8))
        const traceMessage = this._getFunctionTrace(level, trace)
        const stackMessage = stack ? `\n${stack}` : ''

        const coloredOutput = colors.message
            ? chalk.hex(colors.message)(output)
            : output
        const coloredTrace = colors.trace
            ? chalk.hex(colors.trace)(traceMessage)
            : traceMessage
        const coloredStack = colors.stack
            ? chalk.hex(colors.stack)(stackMessage)
            : stackMessage

        const finalMessage =
            chalk.hex(colors.timestamp)(timestamp) +
            chalk.hex(colors.traceId)(formattedTraceId) +
            formattedLogLevel +
            chalk.hex(colors.label)(labelMessage) +
            coloredOutput +
            coloredTrace +
            coloredStack +
            '\n'

        return finalMessage
    }

    private _getTraceId(message: any): string | undefined {
        try {
            const currentSpan = trace.getSpan(context.active())

            if (!currentSpan) return

            const spanContext = trace.getSpan(context.active()).spanContext()

            const parsedMessage = isPlainObject(message)
                ? JSON.stringify(message, (key, value) =>
                      typeof value === 'bigint' ? value.toString() : value,
                  )
                : (message as string)

            currentSpan.addEvent(parsedMessage)

            return spanContext.traceId
        } catch (error) {
            return
        }
    }

    private _findOptions(args: unknown[]): LogMethodOptions {
        const options = args.find(
            (arg) =>
                typeof arg === 'object' && ('trace' in arg || 'label' in arg),
        )

        if (!options) {
            return {}
        }

        return options
    }

    private _findOptionsAndStack(args: unknown[]): {
        options: LogMethodOptions
        stack?: string
    } {
        const options = args.find(
            (arg) =>
                typeof arg === 'object' && ('trace' in arg || 'label' in arg),
        ) as LogMethodOptions

        const stacktraceRegex = new RegExp(/at /g)

        const stack = args.find(
            (arg) => typeof arg === 'string' && stacktraceRegex.test(arg),
        ) as string

        if (!options) {
            return {
                stack,
                options: {},
            }
        }

        return {
            options,
            stack,
        }
    }

    private _isLabelAllowed(label: string): boolean {
        const { labels } = this._config

        if (labels === undefined) {
            return true
        }

        if (labels === false) {
            return false
        }

        return isMatch(label, labels)
    }

    private _getFunctionTrace(level: LogLevel, forceEnable?: boolean): string {
        let isTraceEnabled = this._isTraceEnabledByDefault(level)

        if (this._config.trace === false) {
            isTraceEnabled = false
        }

        if (forceEnable !== undefined) {
            isTraceEnabled = forceEnable
        }

        if (isProduction) {
            isTraceEnabled = false
        }

        if (isTraceEnabled) {
            try {
                throw new Error()
            } catch (error) {
                const stacktrace = error.stack.split(' at ') as string[]
                const context = stacktrace[CONTEXT_INDEX_IN_STACK]
                const [callerFunction, functionTrace] = context.split('(')

                const result: FunctionTrace = {}

                if (!functionTrace) {
                    result.path = callerFunction
                    result.caller = `anonymous function`
                } else {
                    const [path] = functionTrace.split(')')

                    result.path = path
                    result.caller = callerFunction
                }

                const { caller, path } = result

                return `${`\ncaller ${chalk.gray(
                    `->`,
                )} ${caller}`}${`\npath ${chalk.gray(`->`)} ${path}`}`
            }
        }

        return ''
    }

    private _isTraceEnabledByDefault(level: LogLevel): boolean {
        switch (level) {
            case 'warn':
            case 'error':
                return !isProduction

            case 'debug':
                return true

            default:
                return false
        }
    }

    private _getColorByLogLevel(level: LogLevel): (text: string) => string {
        switch (level) {
            case 'debug':
                return clc.magentaBright
            case 'warn':
                return clc.yellow
            case 'error':
                return clc.red
            case 'verbose':
                return clc.cyanBright
            case 'timer':
                return clc.lightGreen
            default:
                return clc.green
        }
    }
}
